Fix: Updating IP address to null doesn't update

achowdhry-ripple opened the pull request:
## High Level Overview of Change

<!--
Please include a summary/list of the changes.
If too broad, please consider splitting into multiple PRs.
If a relevant Asana task, please link it here.
-->

If a user tries to update an existing crawls DB entry with a non-null IP to have a null IP, the IP address does not update. This is because we use merge in our saveNode, which has the behavior of excluding undefined values during conflict resolution. As a result, we need to manually override with the null fields we would like to retain (essentially sanitize the new entry with nulls instead of undefined)

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Before / After

<!--
If just refactoring / back-end changes, this can be just an in-English description of the change at a technical level.
If a UI change, screenshots should be included.
-->

## Test Plan

<!--
Please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

pdp2121 commented:
Can you fix the test case? Should only overwrite if either ip or port is null

mvadari commented:
nit: the PR title should explain what it's fixing, not the name of the issue (it doesn't really make grammatical sense right now)

achowdhry-ripple commented:
@mvadari sorry, had prematurely made this PR to see the pipeline output, should have marked it as draft. It is ready now

ckeshava commented on .vscode/settings.json:
```diff
@@ -25,7 +25,7 @@
         "enable": true
     },
     "editor.codeActionsOnSave": {
-        "source.fixAll.eslint": true
```
Did you mean to exclude this update from the changelog? I'd prefer the linter run automatically, what do you think?

achowdhry-ripple replied:
The line right below this maintains that the litner runs automatically. As of the latest VSCode versions, this boolean setting is converted to "explicit" so it was bound to be changed at some point. Seems like it is okay to keep as long as people are updating their vs code to the latest version

https://stackoverflow.com/questions/77637621/vscode-workspace-settings-change-on-its-own

ckeshava commented on src/crawler/crawl.ts:
```diff
@@ -233,7 +233,7 @@ class Crawler {
 
       this.publicKeysSeen.add(normalizedPublicKey)
-      promises.push(saveNode(dbNode))
+      promises.push(saveNode({ ip: undefined, port: undefined, ...dbNode }))
```
why are the `ip` and `port` values set to `undefined`? The `node` object already contains the correct `ip`, `port` fields, as returned from the `Crawl.crawlNode` method.

achowdhry-ripple replied:
This is a workaround. Essentially, the database knex operation "merge" used in saveNode does not automatically set unincluded fields to null. When crawling, if an IP is null, it is currently excluded as a field from the JSON object, which then results in the new ip = null not being updated into the database. With this logic, if dbNode has an ip or port it will overwrite the undefined ones I have set since they occur later. However if it does not (as related to the bug this PR solves specifically), ip and port will still appear as undefined fields explicitly rather than just not being included

ckeshava replied:
If a server operator wants to hide their ip/port information, should we retain that information in the crawls table? In this case, why shouldn't we delete the row pertaining to this node's `public_key` ?

Without the ip address and port information, this row in the `crawls` is not useful. All the other columns of the crawls table (like `ws_url`, `server_state`, etc) are not useful because the said server is unreachable.

pdp2121 replied:
I think it would just mean that the server becomes private. ip is not the only useful information for crawls table since we also show other information on the Explorer Nodes page. It's just not gonna be used for connections

ckeshava replied:
<img width="884" alt="image" src="https://github.com/user-attachments/assets/84ab9212-3e93-4067-86a0-b20fdd09c642" />
Without the `ip:port` address, all of the other information is irrelevant, because the other columns change with time. If the node is not queried periodically, then all the info listed in the Explorer's Nodes page will be stale.

Schema of the `crawls` table: https://github.com/ripple/validator-history-service/blob/main/ARCHITECTURE.md#crawls


<img width="1891" alt="image" src="https://github.com/user-attachments/assets/f7087ad6-2a40-483e-9f74-be55392b0e42" />

> I think it would just mean that the server becomes private.

I understand that node-operators choose to use `peer_private` in their rippled configurations.

These are the other places where the `crawls` table is used, however the `validators`, `networks` and `manifests` tables are more authoritative sources of truths.

1. [getNetworkFromPublicKey](https://github.com/ripple/validator-history-service/blob/51d5db952bb7a776651e7bb67a5fe416195b6d6e/src/api/routes/v1/get-network.ts#L176): The association of `public_key` with `networks` is periodically refreshed in the `validators` table.
2. [handleHealth](https://github.com/ripple/validator-history-service/blob/51d5db952bb7a776651e7bb67a5fe416195b6d6e/src/api/routes/v1/health.ts#L16): This is not a useful API endpoint. The connected column of the crawls table is [not populated correctly](https://github.com/ripple/validator-history-service/issues/323).
3. [findInDatabase](https://github.com/ripple/validator-history-service/blob/51d5db952bb7a776651e7bb67a5fe416195b6d6e/src/api/routes/v1/nodes.ts#L64) and [getNodes](https://github.com/ripple/validator-history-service/blob/51d5db952bb7a776651e7bb67a5fe416195b6d6e/src/api/routes/v1/nodes.ts#L109) : We need to use an outer-join of `validators` (or) `manifest` table with the `location` table here.

The rest of the uses of the `crawls` table are all utilities for db operations, `crawl` (or) `connection`-management operations.

ckeshava replied:
If this is out of scope of the current PR, I am happy to implement in a different PR. What do you guys think?

pdp2121 replied:
I dont think this is true `Without the `ip:port` address, all of the other information is irrelevant, because the other columns change with time. If the node is not queried periodically, then all the info listed in the Explorer's Nodes page will be stale.` Private node still share information with its peer, so we can still get its information updated from its peer. By being private only means that 1. we cannot explore its peer, 2. we cannot get a websocket connection. The remaining content please raise an issue or create a PR, I will be happy to review

ckeshava replied:
Suppose `A` is behind a firewall and it has configured `B` as its `private_peer`. Hence, `A` is behind a firewall, whereas `B` is accessible by the world.

> Private node still share information with its peer

I'm confused because this contradicts the documentation. The [highlighted section](https://xrpl.org/docs/references/http-websocket-apis/peer-port-methods/peer-crawler#response-format) of the documentation states that if a node is configured as a `private_peer`, then its information is not included in the `peer_crawler` API response. So, `A` will not share information about its (say) forward proxies (`B`).

Are you suggesting the other direction of communication? If `B` shares information about `A`, doesn't it remove the "privacy" aspect of `A` ? Am I missing something here?

<img width="882" alt="image" src="https://github.com/user-attachments/assets/88fcd6fe-d401-4039-a10a-0f266b20f8fd" />

achowdhry-ripple replied:
This conversation now seems out of scope for this PR. Since this PR specifically addresses a customer request, @ckeshava could you move this conversation to a new issue/PR?

ckeshava replied:
@achowdhry-ripple  I still feel that a better solution (for this issue) is to remove the entry from the `crawls` table, instead of setting the `ip`/`port` information to `NULL`.

I don't mean to stall this PR, I'll watch out for any edge-cases that are not covered by my proposal.

pdp2121 replied:
The highlighted section only meant that the ip field will be omitted if the peer is private. Other info will still be available. So not sure why we should remove the row, the client that asked for this only asking for ip/port to be removed when their node became private, which likely means other info would still be available

achowdhry-ripple replied:
closing due to urgency and since this is nonbreaking. @ckeshava feel free to track this conversation in a new issue that can look into cleanliness of deleting a row vs updating to null

ckeshava commented on test/crawler/crawler.test.ts:
```diff
@@ -65,6 +66,31 @@ describe('Runs test crawl', () => {
+  test('successfully updates ip and port to null', async () => {
+    await crawl('1.1.1.1')
+
+    Object.keys(network3.peers).forEach((peer: string) => {
+      nock(`https://${peer}:51235`)
+        .get('/crawl')
+        .reply(
+          200,
+          (network3.peers as Record<string, Record<string, unknown>>)[peer],
+        )
+    })
```
This snippet can be added into the `mock()` function on line 10. I prefer this because there is a `beforeAll()` function which invokes `mock()`. This could ease future development of test cases.

achowdhry-ripple replied:
I cannot do that because I am changing the simulated nodes in network1 specifically for this test, as opposed to creating an actual 3rd network that all tests would use. The code you have highlighted iterated over network3 peers, which all map to the same IP addresses as network 1 so I am updating the nock endpoints with the new node values, which includes the newly null ip node

ckeshava commented on test/crawler/crawler.test.ts:
```diff
@@ -65,6 +66,31 @@ describe('Runs test crawl', () => {
   })
 
+  test('successfully updates ip and port to null', async () => {
+    await crawl('1.1.1.1')
```
```suggestion
```

Invoking the crawl method once is sufficient.

ckeshava commented on test/crawler/fixtures/null-ip-three-node-crawl.json:
```diff
@@ -0,0 +1,94 @@
+      "overlay": {
+        "active": [
+          {
+            "ip": null,
```
In the broader context of this PR, what does this test represent? If a peer designates the ip, port information of an active_node as `null`, then we should be able to save it into the data base. Is that correct?

achowdhry-ripple replied:
yes, this designates the response we would get when crawling and a peer was changed to a null IP and port value.

pdp2121 replied:
I would prefer ip and port here to not be set instead of null to align with behavior described in Peer Crawler doc

Patel-Raj replied:
@achowdhry-ripple The unit tests fail in both of these cases:

1. comment out the test named `successfully crawls 3 node network`
2. Remove ip and port fields altogether from null-ip-three-node-crawl.json file for 1.1.1.1 node.

ckeshava replied:
@Patel-Raj 
> comment out the test named successfully crawls 3 node network

I had a similar observation. This occurs due to the set up of `nock` endpoints. If the first test is removed, then the set up in `beforeAll` is not conducive for the current test. 

This is not a problem with this code change. It appears to be an artifact of the additional `nock` endpoint on the URL.

achowdhry-ripple replied:
this has now been fixed, and I separated the nocks per test for clarity.

ckeshava reviewed (changes requested):
This PR appears to contradict this [line of code](https://github.com/ripple/validator-history-service/blob/910744a4e39946f3ef0dbe83fed13e99cca1af41/src/crawler/crawl.ts#L238)

If the ip (or) the port information is `null`, we do not want to crawl that endpoint (or) save it into the data-base. Is my understanding correct? Can you give me some examples for exceptions to this idea?

ckeshava commented:
Here is the documentation page for the appropriate [peer_crawler request](https://xrpl.org/docs/references/http-websocket-apis/peer-port-methods/peer-crawler#response-format). The `ip` and `port` fields might be omitted from the response, but I don't expect them to be `null`/`undefined`. Are you planning to handle the former case?
<img width="880" alt="image" src="https://github.com/user-attachments/assets/57344838-afa5-4558-9684-8f993d429e18" />

achowdhry-ripple commented:
> Here is the documentation page for the appropriate [peer_crawler request](https://xrpl.org/docs/references/http-websocket-apis/peer-port-methods/peer-crawler#response-format). The `ip` and `port` fields might be omitted from the response, but I don't expect them to be `null`/`undefined`. Are you planning to handle the former case? <img alt="image" width="880" src="https://private-user-images.githubusercontent.com/21219765/429689204-57344838-afa5-4558-9684-8f993d429e18.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM2MzQwMzYsIm5iZiI6MTc0MzYzMzczNiwicGF0aCI6Ii8yMTIxOTc2NS80Mjk2ODkyMDQtNTczNDQ4MzgtYWZhNS00NTU4LTk2ODQtOGY5OTNkNDI5ZTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDIyNDIxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzOTNkNmY3YmZkMTYwMGVhM2E5ZDhlZDUzOWY1Y2IxOGFkYjYxNGIyYjc2NmZhNzQ2ZGRhNjIwMzNhYWJlYTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gyiiyga-WpWxD6DK9V2uZt-YSEAqn2bdDAew1X944FU">

This PR fixes a bug with the former case. If a node changes from a non-null IP to a null IP, we want that change reflected in the DB. Currently, it doesn't update, which is why this issue was raised from a client

achowdhry-ripple commented:
@ckeshava I see that without the context this change can seem obscure, so I have changed some variable names and added comments to the test for additional clarity

pdp2121 commented:
LGTM outside of the small change to test case mentioned

pdp2121 commented:
Also, might be worth reach out to Brett to know which node has become private to make sure that would be fixed with this PR

ckeshava commented on src/crawler/crawl.ts:
```diff
@@ -233,7 +233,7 @@ class Crawler {
       }
 
       this.publicKeysSeen.add(normalizedPublicKey)
```
Shouldn't this be gated by an `if` condition? How do you ensure that you are not over-writing the valid ip/port endpoints with the NULL values? (I'm referring to those server operators who still want to be discovered by the community)

ckeshava replied:
Oh wait, since you have defined these fields `ip: undefined, port: undefined` in the beginning of the object, they can be over-written with valid values, if they exist. that makes sense

achowdhry-ripple commented:
Team, have fixed the issues. First I separated the nocks into each test in crawl.test.ts, rather than have all of them in beforeAll. this made the behavior easier to follow since they are declared as needed. Next, I added a param to the saveNode method since when crawling this_node, if there is not an IP passed in so we do not want to overwrite to null (we assume it was retrieved from a peer). In the peer iterations, we do want to set to null

pdp2121 commented on src/shared/database/index.ts:
```diff
@@ -49,17 +49,25 @@ export async function getNetworks(): Promise<Network[]> {
     node.complete_ledgers = ledgersSplit[ledgersSplit.length - 1]
   }
+
+  const sanitizedMergeNode = shouldOverwriteNull
```
Maybe we can just set ip and port to null in the dbNode in `promises.push(saveNode(dbNode, true))` above? Then we don't need to modify saveNode at all?

achowdhry-ripple replied:
I had it this way at first but it didn't work because then it always set the this_node IP to null, even though it might have already had an IP in a previous crawl from a different node.This resulted in a lot of nodes with null IP and port even though it wasn't actually the case. Thats why I added the variable to overwrite or not. Also IP and port do not allow for null by typing, so it would make the typing a bit more messy

Patel-Raj commented on src/crawler/crawl.ts:
```diff
@@ -233,7 +234,9 @@ class Crawler {
-      promises.push(saveNode(dbNode))
+
+      // If a peer has the node IP as null, we want to overwrite in the database
+      promises.push(saveNode(dbNode, true))
```
@achowdhry-ripple Can you please check if only this change will fix the issue as suggested by @pdp2121 as well. 
We can remove boolean from the saveNode method and saveNode remains the same as its in main branch?

```suggestion
      promises.push(saveNode({ ip: null, port: null, ...dbNode}))
```

achowdhry-ripple replied:
@Patel-Raj This would ruin the typing for the Node object, and require me to change `ip` and `port` to allow for null which breaks other parts of the code for me that reference these parameters. Personally I found this more messy and would rather update `saveNode` than make the typing here more opaque. Can discuss further if you feel strongly about this, or if I missed something and you were able to make this change a different way that doesn't require typing changes?

Patel-Raj replied:
@achowdhry-ripple Ya, I had to change the type for ip and port and add a null check just below this line.

`promises.push(saveNode({ ip: null, port: null, ...dbNode }))`

```
ip?: string | null
port?: number | null
```

You are right, this also, gets a little messy but keeps the saveNode clean as its purpose is to save or update the node if it exists. No, strong opinions.

achowdhry-ripple replied:
true, I'd rather keep it like this for typing clarity in general, since this essentially adds a toggle for modifying the merge functionality in the db save which could be extensible in the future if there are more asks like this (other fields that should be nullable). If anyone strongly disagrees and would rather have it the other way, am open to discuss but will otherwise close out this thread and merge soon
